### PR TITLE
Redo DateTimeInput with downshift

### PR DESCRIPTION
**Pull request by dferber90**
This is the new  `DateTimeInput` based on `downshift`.

| Before | After |
| ----- | ---- |
| image | image |

See also #281 which replaces `DateInput`.

**Breaking Change** Previously `onChange` would be called with the value directly. Now it gets called with an event containing the value. This was done to be in line with how the other inputs work.

### CODE REVIEW

File: src/components/inputs/date-time-input/README.md [hunk 1]
```diff
@@ -8,32 +8,37 @@ import { DateTimeInput } from '@commercetools-frontend/ui-kit';
...
+| `placeholder`          | `string` |    -     | -                                  | -       | Placeholder value to show in the input field                                                                                            |
+| `isDisabled`           | `bool`   |    -     | -                                  | `false` | Disables the date picker                                                                                                                |
+| `horizontalConstraint` | `object` |    -     | `xs`, `s`, `m`, `l`, `xl`, `scale` | `scale` | Horizontal size limit of the input field.                                                                                               |
+
+##### `props.timeZone`
+
+The `timeZone` specifies which time zone the calendar uses, which time zone the selected value is shown in and which time zone entered dates and times are parsed in. For example, given the value is `2018-10-18T00:00:00.000Z and the`timeZone`is set to`Europe/Berlin`, then entering "13:00" as the time results in the`onChange`being called with`2018-10-15T11:00:00.000Z` (as the time zone offset for that date is two hours).
```
Comment (montezume) [inline src/components/inputs/date-time-input/README.md]: I find this explanation quite confusing. Maybe we can rewrite it together?

File: src/components/inputs/date-time-input/README.md [hunk 2]
```diff
@@ -8,32 +8,37 @@ import { DateTimeInput } from '@commercetools-frontend/ui-kit';
...
+
+> Note: It is not possible to supply a general time-zone offset, one has to always use the time-zone. This is because the offset depends on the selected date and time. That's why it's also not possible to get the general offset of a time zone from UTC. It is only possible to get the offset from UTC for a specific date & time.
```
Comment (montezume) [inline src/components/inputs/date-time-input/README.md]: can we either use timeZone (the prop), or `time zone`, (the english word)?

File: src/utils/calendar-time.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
+// Difference of this file to utils.js is that this file
```
Comment (montezume) [inline src/utils/calendar-time.js]: why the name `calendar-time`?

Comment (dferber90) [reply src/utils/calendar-time.js]: Those are utilities for a calendar which respects time. It was really hard to find good names here. We need the utils to:
- share logic between the calendars (`DateInput` &  `DateRangInput`). `DateTimeInput` uses its own utils (this file).
- keep the date library (`moment`) out of our components which makes them more readable and makes it easier to replace the date-library if we ever want to

File: src/components/inputs/date-time-input/date-time-input.js [hunk 4]
```diff
@@ -1,350 +1,424 @@
...
-  }
+const activationTypes = [
+  Downshift.stateChangeTypes.keyDownEnter,
+  Downshift.stateChangeTypes.clickItem,
+];
 
-  return addFlatpickrOffset(value, timeZone);
+const preventDownshiftDefault = event => {
+  // eslint-disable-next-line no-param-reassign
+  event.nativeEvent.preventDownshiftDefault = true;
```
Comment (tdeekens) [inline src/components/inputs/date-time-input/date-time-input.js]: That's a really nice way to give a lot of opt-in control to consumers.

File: src/components/inputs/date-time-input/date-time-input.js [hunk 5]
```diff
@@ -1,350 +1,424 @@
...
+    const today = getToday(this.props.timeZone);
+    this.setState(
+      prevState => ({
+        calendarDate: today,
+        highlightedIndex:
+          prevState.suggestedItems.length +
+          getDaysInMonth(today, this.props.timeZone) -
+          1,
+      }),
+      () => this.inputRef.current.focus()
```
Comment (tdeekens) [inline src/components/inputs/date-time-input/date-time-input.js]: 💯

File: src/components/inputs/date-time-input/date-time-input.js [hunk 6]
```diff
@@ -1,350 +1,424 @@
...
+                      )}
+                      yearLabel={getYearCalendarLabel(
+                        this.state.calendarDate,
+                        this.props.intl.locale
+                      )}
+                      onPrevMonthClick={() => this.jumpMonth(-1)}
+                      onTodayClick={this.showToday}
+                      onNextMonthClick={() => this.jumpMonth(1)}
+                      onPrevYearClick={() => this.jumpMonth(-12)}
+                      onNextYearClick={() => this.jumpMonth(12)}
```
Comment (tdeekens) [inline src/components/inputs/date-time-input/date-time-input.js]: Nitpick: `jumpMonths`? We literally jump multiple?

**Comment by emmenko** [comment]
Nitpick: can the default day be highlighted a bit more? It's a bit "hard" to see the blue color. Also the months/years do not seem to be aligned correctly.

image

**Comment by dferber90** [comment]
> can the default day be highlighted a bit more? It's a bit "hard" to see the blue color.

The blue color represents "today". It is not the default highlighted date.A highlight looks like [this](https://user-images.githubusercontent.com/1765075/49505530-427e3280-f87c-11e8-82b3-67d47a0b947d.png). The 14th is highlighted here.


> Also the months/years do not seem to be aligned correctly.

Thanks for the hint. This was a problem in Firefox only. I fixed it there now.

**Pull request merged**